pkg/kv/kvserver/replica_proposal.go

```diff
@@ -707,21 +709,35 @@ func (r *Replica) handleReadWriteLocalEvalResult(ctx context.Context, lResult re
 		lResult.MaybeAddToSplitQueue = false
 	}
 
+	// The following three triggers require the raftMu to be held. If a
+	// trigger is present, acquire the mutex if it is not held already.
+	maybeAcquireRaftMu := func() func() {
+		if withRaftMu {
+			return func() {}
+		}
+		withRaftMu = true
+		r.raftMu.Lock()
+		return r.raftMu.Unlock
+	}
+
 	if lResult.MaybeGossipSystemConfig {
+		defer maybeAcquireRaftMu()()
 		if err := r.MaybeGossipSystemConfigRaftMuLocked(ctx); err != nil {
```

Comment:
While you're here, sprinkle `raftMu.AssertHeld()` into the three methods with the `RaftMuLocked` suffix?

Reply:
It's already there. That's the cause of the test flakiness 😃